README.md

```diff
@@ -1,73 +1,320 @@
+
+#### public_key
+
+`public_key_src` specifies a path to a file on the controller host containing the public key used by this host for public key authentication without certificates. Otherwise, the user can directly specify the public key for this host by populating `public_key_content`. `public_key_content` can also accept a CKAID or nickname for a public key in the NSS database.
+
+Note that `public_key_src` and `public_key_content` may also be specified as host variables in the Ansible `all.hosts` section. The variable names in this case will be `vpn_public_key_src` and `vpn_public_key_content`. See [this example](#defining-public-keys-at-different-levels).
+
+**To be added later**: if neither `public_key_src` nor `public_key_content` is populated, the role will generate key pairs for each host.
+
+### Handling special cases with host names
+
+There are two ways to handle the special case when the names of the hosts in the Ansible inventory are not FQDN hostnames that you can use directly (e.g. they may be host aliases like "bastion_east"):  
+
+1. **Use the Ansible `hosts` configuration**
 
 ```yaml
-template_foo: "oof"
-template_bar: "baz"
+all:
+  hosts:
+    bastion_east:
+      ansible_host: bastion1.example.com
+      vpn_hostid: '@bastion-east'
+    bastion_west:
+      ansible_host: bastion2.example.com
+      vpn_hostid: '@bastion-west'
+    bastion_north:
+      ansible_host: bastion3.example.com
+      vpn_hostid: '@bastion-north'
+  vars:
+    vpn_connections:
+      - hosts:
+          bastion_east:
+          bastion_west:
+          bastion_north:
 ```
+The role would lookup the value of the host to use from `hostvars[hostname].ansible_host`.  You can specify other variables which are always associated with a particular host, like `vpn_hostid`, in the host configuration, which would be used for every connection where that host is referenced.
 
-### Variables Exported by the Role
+2. **Directly specify the host to use**
 
-This section is optional.  Some roles may export variables for playbooks to
-use later.  These are analogous to "return values" in Ansible modules.  For
-example, if a role performs some action that will require a system reboot, but
-the user wants to defer the reboot, the role might set a variable like
-`template_reboot_needed: true` that the playbook can use to reboot at a more
-convenient time.
+Users can provide the host specific parameters with either the host's definition in the Ansible `all.hosts` section, or under each host in the `hosts` list under `vpn_connections`.  This gives the admin flexibility in case they cannot edit values in one or the other section.
 
-Example:
+```yaml
+all:
+  hosts:
+    bastion_east:
+      ansible_host: bastion1.example.com # the hostname that Ansible uses
+      vpn_host: 192.168.122.101 # the IP address we want to use for the tunnel
```

Comment:
Is it a general rule that in order to set a in the inventory a parameter that belongs under a host in `vpn_connections[...].hosts`, you prefix the parameter with `vpn_`? So to set `public_key_src`, you would use `vpn_public_key_src` in the inventory, etc.

Reply:
Technically, all of those parameters are keys/nested keys within the `vpn_connections` dictionary, so I don't think the `vpn_` convention applies.

Reply:
Right.  If you are setting a vpn parameter in the context of the host definition in the inventory, then you must use a `vpn_` prefix because that is a "global" namespace.  However, the same parameter in the context of a tunnel or host definition within the `vpn_connections` list does not need the `vpn_` prefix - it is implied.